File: gjsgl/frontend.py
#a Imports
import glfw
import traceback
from OpenGL import GL
from typing import *

#a Classes
#c Frontend
class Frontend:
    #v Properties
    window_title : ClassVar[str] = "blah"
    animating   : bool
    main_window : int
    frame_delay : float = 0.02
    keys_down   : Set[int]
    key_mods    : int # bitmask of shift, ctrl, alt, meta keys down (bits 0 to 3)
    time        : float
    time_last   : float
    mouse_pos   : Tuple[float,float]
    mouse_pos_drag_start   : Tuple[float,float]
    # touches : Dict[TouchId, Any] - in JS, for touch Web interface
    #f __init__
    def __init__(self) -> None:
        glfw.init()
        glfw.window_hint(glfw.CONTEXT_VERSION_MAJOR, 4)
        glfw.window_hint(glfw.CONTEXT_VERSION_MINOR, 1)
        glfw.window_hint(glfw.OPENGL_FORWARD_COMPAT, GL.GL_TRUE)
        glfw.window_hint(glfw.OPENGL_PROFILE, glfw.OPENGL_CORE_PROFILE)
        self.main_window = glfw.create_window(720, 600, "Opengl GLFW Window", None, None)
        glfw.set_key_callback(self.main_window, self.key)
        glfw.set_cursor_pos_callback(self.main_window, self.cursor_pos)
        glfw.set_mouse_button_callback(self.main_window, self.mouse_button)
        glfw.make_context_current(self.main_window)
        for x in [GL.GL_VENDOR , GL.GL_RENDERER , GL.GL_VERSION , GL.GL_SHADING_LANGUAGE_VERSION]:
            print(GL.glGetString(x))
            pass

        self.animating = False
        self.time_last = glfw.get_time()
        self.keys_down = set()
        self.key_mods = 0
        self.buttons_down = 0
        self.mouse_pos_drag_start = (0.,0.)
        self.mouse_pos = (0.,0.)
        pass
    #f swap_buffers
    def swap_buffers(self) -> None:
        """
        swap_buffers must be called on completion of a framebuffer update
        It is not required in the Web library as its framebuffer is managed by the canvas
        """
        glfw.swap_buffers(self.main_window)
        pass
    #f set_animating
    def set_animating(self, animating:bool) -> None:
        """
        Set animating - in JS it kicks of the animating cycle; in Python it is not yet clear
        """
        self.animating = animating
        pass
    #f run
    def run(self) -> None:
        """
        Invoked  to run the application, after it has been constructed.
        """
        self.gl_ready()
        self.set_animating(True)
        try:
            while self.animating and not glfw.window_should_close(self.main_window):
                time_now = glfw.get_time()
                timeout = self.time_last + self.frame_delay - time_now
                if timeout > 0:
                    glfw.wait_events_timeout(timeout)
                    pass
                else:
                    glfw.poll_events()
                    pass
                time_now = glfw.get_time()
                if time_now > self.time_last + self.frame_delay:
                    self.time_last = time_now
                    glfw.make_context_current(self.main_window)
                    self.handle_tick(time_now, self.time_last)
                    pass
                pass
            pass
        except Exception as e:
            print(f"Failed: {e}")
            print(f"Failed: {traceback.format_exc()}")
            pass
        glfw.terminate()
        pass
    #f key
    def key(self, window:Any, key:int, scancode:int, action:int, mods:int) -> None:
        """
        Callback invoked on a key press/release/repeat
        Invoke the subclass action 'key_fn' that can be common between JS and Python
        """
        press = (action==glfw.PRESS) or (action==glfw.REPEAT)
        if ((key==81) and (mods&2) and press): self.set_animating(not self.animating)
        if press:
            self.keys_down.add(key)
            pass
        else:
            self.keys_down.discard(key)
            pass
        self.key_fn(key, scancode, press, mods)
        pass
    #f cursor_pos
    def cursor_pos(self, window:Any, xpos:float, ypos:float) -> None:
        """
        Callback invoked on a mouse position event
        Invoke the subclass 'cursor_fn'
        """
        self.mouse_pos = (xpos,ypos)
        self.cursor_fn(xpos, ypos)
        pass
    #f mouse_button
    def mouse_button(self, window:Any, button:int, action:int, mods:int) -> None:
        """
        Callback invoked on a mouse button event
        Invoke the subclass 'mouse_fn'
        """
        (xpos, ypos) = glfw.get_cursor_pos(window)
        if action==glfw.PRESS:
            self.buttons_down |= 1<<button
            self.mouse_pos_drag_start = (xpos, ypos)
            pass
        elif action==glfw.RELEASE:
            self.buttons_down &= ~(1<<button)
            pass
        self.key_mods = mods
        self.mouse_button_fn(xpos, ypos, button, action, mods)
        pass
    #f key_fn
    def key_fn(self, key:int, scancode:int, press:bool, mods:int) -> None:
        """
        Method to be subclassed - same in JS and Python
        """
        print(f"{key} {scancode} {press} {mods}")
        pass
    #f cursor_fn
    def cursor_fn(self, xpos:float, ypos:float) -> None:
        """
        Method to be subclassed - same in JS and Python
        """
        print(f"{self.mouse_pos}, {self.mouse_pos_drag_start} {self.buttons_down} {self.key_mods}")
        pass
    #f mouse_button_fn
    def mouse_button_fn(self, xpos:float, ypos:float, button:int, action:int, mods:int) -> None:
        """
        Method to be subclassed - same in JS and Python
        """
        print(f"{self.mouse_pos}, {self.mouse_pos_drag_start} {self.buttons_down} {self.key_mods} {button} {action} {mods}")
        pass
    #f All done
    pass

File: gjsgl/frontend.js
//c Frontend
var GL;
class Frontend {
    constructor() {
        const canvas = document.querySelector("#glcanvas");
        const gl     = canvas.getContext("webgl2");
        if (!gl) { alert("Unable to start webgl"); }
        GL = gl;
        document.addEventListener("keydown",     (k) => this.key(k) );
        document.addEventListener("keyup",       (k) => this.key(k) );
        canvas.addEventListener("mousedown", (m) => this.mouse(m) );
        canvas.addEventListener("mouseup",   (m) => this.mouse(m) );
        canvas.addEventListener("mousemove", (m) => this.mouse(m) );
        canvas.addEventListener("touchstart",  (t) => this.touch(t), false);
        canvas.addEventListener("touchend",    (t) => this.touch(t), false);
        canvas.addEventListener("touchcancel", (t) => this.touch(t), false);
        canvas.addEventListener("touchmove",   (t) => this.touch(t), false);

        this.animating = false;
        this.run_step_pending = false;
        this.time_last = 0.;
        this.time = 0.;
        this.keys_down = new Set();
        this.key_mods = 0;
        this.buttons_down = 0;
        this.mouse_pos_drag_start = [0.,0.];
        this.mouse_pos = [0.,0.];
        this.touches = new Map();
    }
    //f run
    run() {
        const promise = this.init();
        promise.then( (x) => {
            console.log("Frontend initialized");
            this.gl_ready();
            console.log("Frontend GL ready");
            this.set_animating(true);
        } );
    }
    //f gl_ready
    gl_ready() {
        console.log("Do GL ready stuff");
    }
    //f set_animating
    set_animating(a) {
        console.log("Set animating",a);
        if (a) {
            if (this.run_step_pending) {return;}
            this.animating = true;
            this.init_time = Date.now();
            this.run_step();
        } else {
            this.animating = false;
        }
    }
    //f run_step
    run_step() {
        this.run_step_pending = false;
        if (this.animating) {
            this.time_last = this.time;
            this.time = (Date.now() - this.init_time) * 0.001;
            this.handle_tick(this.time, this.time_last);
            requestAnimationFrame(()=>this.run_step());
            this.run_step_pending = true;
        }
    }
    //f key
    key(event) {
        const key = event.keyCode;
        const press = (event.type!="keyup");
        const scancode = 0;
        const mods = (event.shiftKey?1:0)  | (event.ctrlKey?2:0) | (event.altKey?4:0);
        this.key_mods = mods;
        if ((key==81) && (mods&2) && press) {this.set_animating(!this.animating);}
        if (press) {
            this.keys_down.add(key);
        } else {
            this.keys_down.delete(key);
        }
        this.key_fn(key, scancode, press, mods);
    }
    //f key_fn
    key_fn(key, scancode, press, mods) {
        console.log(""+key+" "+scancode+" "+press+" "+mods);
    }
    //f mouse
    mouse(event) {
        const xpos = event.offsetX;
        const ypos = event.offsetY;
        const mods = (event.shiftKey?1:0)  | (event.ctrlKey?2:0) | (event.altKey?4:0);
        this.key_mods = mods;
        if (event.type=="mousemove") {
            this.cursor_fn(xpos, ypos);
        } else {
            if (event.type=="mousedown") {
                this.buttons_down |= 1<<event.button;
            } else if (event.type=="mouseup") {
                this.buttons_down &= ~1<<event.button;
            }
            this.mouse_button_fn(xpos, ypos, event.button, event.type, mods);
        }
        this.mouse_pos = [xpos, ypos];
    }
    //f touch
    touch(event) {
        const mods = (event.shiftKey?1:0)  | (event.ctrlKey?2:0) | (event.altKey?4:0);
        if (event.type=="touchstart") {
            event.preventDefault();
            for (const t of event.changedTouches) {
                this.touches.set(t.identifier,this.touch_start(t.identifier, t.pageX, t.pageY, mods));
            }
        } else if (event.type=="touchmove") {
            for (const t of event.changedTouches) {
                const handle = this.touches.get(t.identifier);
                this.touch_fn(handle, t.identifier, t.pageX, t.pageY, 1, mods);
            }
        } else if (event.type=="touchend") {
            for (const t of event.changedTouches) {
                const handle = this.touches.get(t.identifier);
                this.touch_fn(handle, t.identifier, t.pageX, t.pageY, 2, mods);
                this.touches.delete(t.identifier);
            }
        } else if (event.type=="touchcancel") {
            for (const t of event.changedTouches) {
                const handle = this.touches.get(t.identifier);
                this.touch_fn(handle, t.identifier, t.pageX, t.pageY, 3, mods);
                this.touches.delete(t.identifier);
            }
        }
    }
    //f cursor_fn
    cursor_fn(xpos, ypos) {
        console.log(this.mouse_pos+" "+this.mouse_pos_drag_start+" "+this.buttons_down+" "+this.key_mods);
    }
    //f mouse_button_fn
    mouse_button_fn(xpos, ypos, button, action, mods) {
        console.log(this.mouse_pos+" "+this.mouse_pos_drag_start+" "+button+" "+mods);
    }
    //f touch_start
    touch_start(touch_id, xpos, ypos, mods) {
        return touch_id;
    }
    //f touch_fn
    touch_fn(touch_handle, touch_id, xpos, ypos, action, mods) {
    }
    //f All done
}

